_posts/2021-01-05-overlappable-instances.markdown

```diff
@@ -0,0 +1,263 @@
+  get _ = pure Nothing
+
+instance Logger IO where
+  log _ = pure ()
+
+main :: IO ()
+main = program
+```
+This is clearly limiting, because we are allowed to have only one typeclass instance for a given monad. If we would like to have a NoOp instance and a real-world instance, we have to use newtype wrappers. Let's add aliases then for our not-so-useful instances like this:
+
+```haskell
+newtype NoOpStoreT m a = NoOpStoreT { runNoOpStoreT :: m a }
+  deriving newtype (Functor, Applicative, Monad)
+
+newtype NoOpLoggerT m a = NoOpLoggerT { runNoOpLoggerT :: m a }
+  deriving newtype (Functor, Applicative, Monad)
+
+instance Monad m => Store (NoOpLoggerT m) where
+  put _ _ = pure ()
+  get _ = pure Nothing
+
+instance Monad m => Logger (NoOpLoggerT m) where
+  log _ = pure ()
+
+main :: IO ()
+main = runNoOpStoreT . runNoOpLoggerT $ program
+
+```
+This compiles and runs fine. We need to create an instance of each class for every transformer we intend to use in our stack. Let's see what happens when we reverse the order in which we apply the transformations, that is:
+```haskell
+main :: IO ()
+main = runNoOpLoggerT . runNoOpStoreT $ program
+```
+results in:
+```
+    • No instance for (Logger (NoOpStoreT (NoOpLoggerT IO)))
+        arising from a use of ‘program’
+```
+We've arrived at the infamous O(n^2) instances problem! We can add the missing instances like this:
+```haskell
+instance Monad m => Store (NoOpStoreT m) where
+  put _ _ = pure ()
+  get _ = pure Nothing
+
+instance Monad m => Logger (NoOpStoreT m) where
+  log _ = pure ()
+
+```
+But this gets painful really quickly when we add more classes and transformers.
+Overlappable transformer instances to the rescue! Adding these both for `Store` and `Logger` makes it possible to get rid of boilerplate and allow us to stack our transformers in any* order! Let's start with `Logger`:
+
+```haskell
+instance MonadTrans NoOpStoreT where
+  lift a = NoOpStoreT a
+
+instance (Logger m, MonadTrans t, Monad m) => Logger (t m) where
+  log a = lift $ log a
+
+instance Monad m => Logger (NoOpLoggerT m) where
+  log _ = pure ()
+```
+We've defined a new instance. It is going to be available only if when a `Logger` instance already exists for `m` and we know how to wrap this monad `m` using `t`. For this to be true we need to add a `MonadTrans` instance for our `NoOpStoreT`. Thanks to this we don't have to write all n^2 instances by hand. But the compiler has a problem now:
+
+```
+   • Overlapping instances for Logger (NoOpLoggerT IO)
+        arising from a use of ‘program’
+      Matching instances:
+        instance (Logger m, MonadTrans t, Monad m) => Logger (t m)
+          -- Defined at ...
+        instance Monad m => Logger (NoOpLoggerT m)
+          -- Defined at ...
+
+```
+We need to tell it, that whenever two instances could be used "arising from the use of program", one of them is less specific, and labeled using the `{-# OVERLAPPABLE #-}` pragma. Now the code compiles fine.
```

Comment:
either consider adding more explanation to what "less specific" means OR just link to GHC wiki for folks who want to learn more

Reply:
I think I'll consider this "an exercise for the reader" :-)